File: src/gui/backends/sdl2/widget.rs
use sdl2::surface::Surface;

use crate::gui::Rect;

use super::{Button, Font};

impl Into<sdl2::rect::Rect> for &Rect
{
    fn into(self) -> sdl2::rect::Rect {
        sdl2::rect::Rect::new(
            self.position.0 as i32,
            self.position.1 as i32,
            self.size.0,
            self.size.1,
        )
    }
}

pub enum WidgetType<'a>
{
    Raw(&'a Widget<'a>),
    Button(&'a Button<'a>),
}

impl<'a> WidgetType<'a>
{
    pub fn raw(&self) -> &Widget
    {
        match self
        {
            WidgetType::Raw(wtype) => &wtype,
            WidgetType::Button(wtype) => &wtype.raw(),
        }
    }
}

pub struct Widget<'a>
{
    pub raw: Vec<Rect>,
    pub surfaces: Vec<Surface<'a>>,
    pub font: &'a Font<'a>,
}

impl<'a> Widget<'a>
{
    pub fn new(rectangles: Vec<Rect>, surfaces: Vec<Surface<'a>>, font: &'a Font<'a>) -> Self
    {
        Widget { raw: rectangles, font, surfaces }
    }
}

impl<'a> Widget<'a>
{
    pub fn raw(&self) -> &Widget
    {
        &self
    }

    pub fn from_raw(raw: Widget<'a>) -> Self where Self: Sized {
        raw
    }
}

File: src/gui/backends/sdl2/widget.zig
const sdl2 = @import("sdl2.zig");

const std = @import("std");

const Rect = @import("../../rect.zig");
const button_mod = @import("button.zig");
const font_mod = @import("font.zig");
const window_mod = @import("window.zig");

pub const WidgetType = union(enum)
{
    Raw: *Widget,
    Button: *button_mod.Button,

    const Self = @This();

    pub fn raw(self: *Self) *Widget
    {
        return switch (self.*) {
            inline else => |case| case.raw(),
        };
    }
};

pub const WidgetTexture = struct {
    raw: ?*sdl2.SDL_Texture,
    rect: sdl2.SDL_Rect,
};

pub const WidgetError = error {
    TEXTURE_CREATION_FAILED
};

pub const Widget = struct {
    raw_rects: std.ArrayList(Rect),
    surfaces: std.ArrayList(font_mod.TextSurface),

    const Self = @This();

    pub fn new(rectangles: std.ArrayList(Rect), surfaces: std.ArrayList(font_mod.TextSurface)) Widget
    {
        return Widget {
            .raw_rects = rectangles,
            .surfaces = surfaces,
        };
    }

    pub fn to_textures(self: *Self, window: *window_mod.Window, allocator: std.mem.Allocator) WidgetError!std.ArrayList(WidgetTexture)
    {
        var textures = std.ArrayList(WidgetTexture).init(allocator);

        for (self.surfaces.items) |surface| {
            var texture: ?*sdl2.SDL_Texture = sdl2.SDL_CreateTextureFromSurface(window.renderer, surface.raw);
            if(texture == null)
            {
                std.log.err("SDL_CreateTextureFromSurface Error: {s}\n", .{sdl2.SDL_GetError()});
                return WidgetError.TEXTURE_CREATION_FAILED;
            }

            textures.append(WidgetTexture {.raw = texture, .rect = surface.rect}) catch unreachable;
        }

        return textures;
    }

    pub fn raw(self: *Self) *Widget
    {
        return self;
    }

    pub fn from_raw(widget: Widget) Widget
    {
        return widget;
    }

    pub fn destroy(self: *Self) void
    {
        self.raw_rects.deinit();
        self.surfaces.deinit();
    }
};

test "test widget"
{
}